Token cache

gaguirre opened the pull request:
Implement TokenCache class in KurveIdentity.
Export TokenStorage interface in order to support external token storage provided through IdentitySettings.

Commit:
Implement TokenCache in KurveIdentity to handle expired token renewal. Support external token storage provided through IdentitySettings

Commit:
Remove unecessary cast

Commit:
Revert null assignment to delete operator

Commit:
Remove only matching token for scopes if token is expired instead of all expired tokens

Commit:
Prune expired tokens on TokenCache constructor

Commit:
Type strongly the Token constructor

Commit:
Change TokenStorage signatures for add and remove methods

billba commented:
Okay we satisfied John's comments in Slack. Here are mine:
1. I hate the name of Token, because (a) it’s not a token, and (b) it contains a token and (c) doesn’t tell you its purpose, so:
2. Let’s rename Token -> CachedToken and TokenDictionary => CachedTokenDictionary
3. I think the name of TokenStorage is good. A nice parallel to TokenCache
4. in getForScopes and getForResource (and anywhere else I missed) rename tokens->cachedTokens
5. line 75/76 and line 88/89 turns this into a helper method called deleteCachedToken
6. after 89 you need to return null;
7. 
   scopes.every(scope => cachedToken.scopes.indexOf(scope) >= 0))
   ->
   scopes.every(scope => cachedToken.scopes.some(cachedScope => scope.key == cachedScope.key))
8. inDecodeAccessToken let’s use the typed Token/CachedToken constructor here
9. I’d like to see the declaration for Token/CachedToken look something like this:

```
export class CachedToken {
constructor(
        id: string,
        scopes: string[],
        resource: string,
        token: string,
        expiry: Date
){}
}
```

That way we’re not duplicating all those properties.

I don’t understand the need for any the constructors that take Token and any. Both seem to be making shallow copies. Why not just use the original?

If you do need to make shallow copies, let’s just have constructor(CachedToken). And if you’re calling this with an “any” returned from the TokenStorage service, explicitly cast it to CachedToken. That’s more type safe than an any constructor, which could be randomly called with anything from anywhere. If they’re going to do something stupid, make them work for it.

Commit:
Remove unecessary exports

Commit:
Rename Token by CachedToken

Rename TokenDictionary by CachedTokenDictionary

Rename TokenCache property tokens by cachedTokens

Commit:
Add remove() method for TokenCache and hasScopes() for Token

gaguirre commented:
The above commits should tackle the first 7 items.
Working now in 8 and 9.

Commit:
Refactor CachedToken constructor to a strongly typed one

Commit:
Refactor hasScopes() method
